java/ql/lib/semmle/code/java/security/UnsafeAndroidAccess.qll

```diff
@@ -44,57 +42,75 @@ private class CrossOriginUrlResourceSink extends JavaScriptEnabledUrlResourceSin
  */
 private class JavaScriptEnabledUrlResourceSink extends UrlResourceSink {
   JavaScriptEnabledUrlResourceSink() {
-    exists(Variable settings |
-      webViewLoadUrl(this.asExpr(), settings) and
-      isJSEnabled(settings)
+    exists(WebViewRef webview |
+      isJSEnabled(webview.getAnAccess()) and
+      webViewLoadUrl(this.asExpr(), webview.getAnAccess())
     )
   }
 
   override string getSinkType() { result = "user input vulnerable to XSS attacks" }
 }
 
-/**
- * Holds if a `WebViewLoadUrlMethod` method is called with the given `urlArg` on a
- * WebView with settings stored in `settings`.
- */
-private predicate webViewLoadUrl(Expr urlArg, Variable settings) {
-  exists(MethodAccess loadUrl, Variable webview, MethodAccess getSettings |
-    loadUrl.getArgument(0) = urlArg and
-    loadUrl.getMethod() instanceof WebViewLoadUrlMethod and
-    loadUrl.getQualifier() = webview.getAnAccess() and
-    getSettings.getMethod() instanceof WebViewGetSettingsMethod and
-    webview.getAnAccess() = getSettings.getQualifier() and
-    settings.getAnAssignedValue() = getSettings
-  )
-}
+private class WebViewRef extends Element {
+  WebViewRef() {
+    this.(RefType).getASourceSupertype*() instanceof TypeWebView or
+    this.(Variable).getType().(RefType).getASourceSupertype*() instanceof TypeWebView
+  }
 
-/**
- * A method allowing any-local-file and cross-origin access in the WebSettings class.
- */
-private class CrossOriginAccessMethod extends Method {
-  CrossOriginAccessMethod() {
-    this.getDeclaringType() instanceof TypeWebSettings and
-    this.hasName(["setAllowUniversalAccessFromFileURLs", "setAllowFileAccessFromFileURLs"])
+  /** Gets an access to this WebView. */
+  Expr getAnAccess() {
+    exists(ThisAccess t | t.getType() = this and result = t |
+      t.isOwnInstanceAccess() or
+      t.isEnclosingInstanceAccess(this)
+    )
```

Comment:
This is restricted to explicit `this`, which is relatively rare. Can this be re-expressed to deal in data-flow nodes, where an implicit `this` also has a corresponding node?

Reply:
Good point, done in 8cc5a9f. I guess that makes those predicates less reusable, so I moved them back to `UnsafeAndroidAccess.qll` in bbb98c3.